src/source/test_waveform.rs

```diff
@@ -0,0 +1,196 @@
+//! Generator sources for various periodic test waveforms.
+//!
+//! This module provides several periodic, deterministic waveforms for testing other sources and
+//! for simple additive sound synthesis. Every source is monoaural and in the codomain `[-1.0f32,
+//! 1.0f32]` 
+//!
+//! # Example
+//!
+//! ```
+//! use rodio::source::{TestWaveform,TestWaveformFunction};
+//!
+//! let tone = TestWaveform::new(cpal::SampleRate(48000), 440.0, TestWaveformFunction::Sine);
+//! ```
+use std::f32::consts::TAU;
+use std::time::Duration;
+
+use super::SeekError;
+use crate::Source;
+
+/// Test waveform functions.
+#[derive(Clone, Debug)]
+pub enum TestWaveformFunction {
+    /// A sinusoidal waveform.
+    Sine,
+    /// A triangle wave.
+    Triangle,
+    /// A square wave, rising edge at t=0.
+    Square,
+    /// A rising swatooth wave.
+    Sawtooth,
+}
+
+impl TestWaveformFunction {
+    /// Create a single sample for the given waveform
+    #[inline]
+    fn render(&self, i: u64, period: f32) -> f32 {
+        let i_div_p: f32 = i as f32 / period;
+
+        match self {
+            Self::Sine => (TAU * i_div_p).sin(),
+            Self::Triangle => 04.0f32 * (i_div_p - (i_div_p + 0.5f32).floor()).abs() - 1f32,
+            Self::Square => {
+                if i_div_p % 1.0f32 < 0.5f32 {
+                    1.0f32
+                } else {
+                    -1.0f32
+                }
+            }
+            Self::Sawtooth => 2.0f32 * (i_div_p - (i_div_p + 0.5f32).floor()),
+        }
+    }
+}
+
+/// An infinite source that produces one of a selection of test waveforms.
+#[derive(Clone, Debug)]
+pub struct TestWaveform {
+    sample_rate: cpal::SampleRate,
+    period: f32,
+    f: TestWaveformFunction,
```

Comment:
f -> function

a single letter is almost always confusing. Is it frequency, a function the fraction?